gpt_index/indices/query/query_runner.py

```diff
@@ -137,63 +153,130 @@ def _get_query_obj(
     ) -> BaseGPTIndexQuery:
         """Get query object."""
         index_struct_type = index_struct.get_type()
-        config = self._get_query_config(index_struct)
+        assert index_struct_type != IndexStructType.COMPOSITE
+        config = self._query_config_map.get(index_struct)
         mode = config.query_mode
 
         from gpt_index.indices.registry import INDEX_STRUT_TYPE_TO_QUERY_MAP
         query_cls = INDEX_STRUT_TYPE_TO_QUERY_MAP[index_struct_type][mode]
-        # if recursive, pass self as query_runner to each individual query
-        query_runner = self
         query_kwargs = self._get_query_kwargs(config)
         query_obj = query_cls(
-            index_struct,
-            **query_kwargs,
-            query_runner=query_runner,
+            index_struct=index_struct,
             docstore=self._docstore,
-            recursive=self._recursive,
-            use_async=self._use_async,
+            **query_kwargs,
         )
 
         return query_obj
 
     def query(
         self,
         query_str_or_bundle: Union[str, QueryBundle],
-        index_struct: IndexStruct,
+        index_id: Optional[str] = None,
     ) -> Response:
         """Run query."""
+        if isinstance(self._index_struct, CompositeIndex):
+            if index_id is None:
+                index_id = self._index_struct.root_id
+            index_struct = self._index_struct.all_index_structs[index_id]
+        else:
+            if index_id is not None:
+                raise ValueError('index_id should be used with composite graph')
+            index_struct = self._index_struct
+
         # NOTE: Currently, query transform is only run once
         # TODO: Consider refactor to support index-specific query transform
+
         # TODO: abstract query transformation loop into a separate class
+        # TODO: support query combiner
+        # query_combiner = self._get_query_combiner(index_struct, query_transform)
 
+        # query transform
         query_transform = self._get_query_transform(index_struct)
-        query_combiner = self._get_query_combiner(index_struct, query_transform)
-
-        query_obj = self._get_query_obj(index_struct)
         if isinstance(query_str_or_bundle, str):
             query_bundle = QueryBundle(
                 query_str=query_str_or_bundle,
                 custom_embedding_strs=[query_str_or_bundle],
             )
         else:
             query_bundle = query_str_or_bundle
+        transform_extra_info = {"index_struct": index_struct}
+        query_bundle = query_transform(
+            query_str_or_bundle, extra_info=transform_extra_info
+        )
+
+        # query
+        query_obj = self._get_query_obj(index_struct)
+        if self._recursive:
+            # call recursively
+            nodes = query_obj.retrieve(query_bundle)
```

Comment:
I see what you're going for.

Most nodes do implement `retrieve` in a pretty standard fashion (e.g. vector index just retrieves top k nodes, keyword table index just retrieves keyword matched nodes). The tree index doesn't though. So in its current form composability wouldn't work for tree indexes.

I do like the decoupling of retrieval and synthesis in general though. I think we could probably consider a rewriting of the tree index logic to decouple retrieval and synthesis a bit (e.g. first "retrieve" the leaf nodes that match the query in the retrieve function, then in the `synthesize` function have a way to synthesize a response from these nodes - it can be a bit more complicated than running response builder over these nodes, since we give the option to "refine" the response with parent node context recursively.

Reply:
Good catch. I'm in alignment that we should fix the tree index since it's an exception.